vault/raft.go

```diff
@@ -0,0 +1,664 @@
+					return err
+				}
+				if err := c.barrier.Unseal(ctx, keys[0]); err != nil {
+					c.logger.Error("raft snapshot restore failed to unseal barrier", "error", err)
+					c.sealInternalWithOptions(false, false)
+					return err
+				}
+				c.logger.Info("done reloading master key using auto seal")
+			}
+		}
+
+		if !c.standby {
+			// Go through a preSeal, postUnseal cycle to clear the rest of
+			// vault's in-memory caches
+			c.logger.Info("restarting system after raft snapshot restore")
+			if err := c.preSeal(); err != nil {
+				c.logger.Error("raft snapshot restore failed preSeal", "error", err)
+				return err
+			}
+			{
+				// If the snapshot was taken while another node was leader we
+				// need to reset the leader information to this node.
+				if err := c.underlyingPhysical.Put(ctx, &physical.Entry{
+					Key:   CoreLockPath,
+					Value: []byte(c.leaderUUID),
+				}); err != nil {
+					c.logger.Error("cluster setup failed", "error", err)
+					return err
+				}
+				// re-advertise our cluster information
+				if err := c.advertiseLeader(ctx, c.leaderUUID, nil); err != nil {
+					c.logger.Error("cluster setup failed", "error", err)
+					return err
+				}
+			}
+			if err := c.postUnseal(ctx, ctxCancel, standardUnsealStrategy{}); err != nil {
+				c.logger.Error("raft snapshot restore failed postUnseal", "error", err)
+				return err
+			}
+		}
+
+		return nil
+	}
+}
+
+func (c *Core) JoinRaftCluster(ctx context.Context, leaderAddr string, tlsConfig *tls.Config, retry bool) (bool, error) {
+	if len(leaderAddr) == 0 {
+		return false, errors.New("No leader address provided")
+	}
+
+	raftStorage, ok := c.underlyingPhysical.(*raft.RaftBackend)
+	if !ok {
+		return false, errors.New("raft storage not configured")
+	}
+
+	if raftStorage.Initialized() {
+		return false, errors.New("raft is alreay initialized")
+	}
+
+	init, err := c.Initialized(ctx)
+	if err != nil {
+		return false, errwrap.Wrapf("failed to check if core is initialized: {{err}}", err)
+	}
+	if init {
+		return false, errwrap.Wrapf("join can't be invoked on an initialized cluster: {{err}}", ErrAlreadyInit)
+	}
+
+	transport := cleanhttp.DefaultPooledTransport()
+	if tlsConfig != nil {
+		transport.TLSClientConfig = tlsConfig.Clone()
+		if err := http2.ConfigureTransport(transport); err != nil {
+			return false, errwrap.Wrapf("failed to configure TLS: {{err}}", err)
+		}
+	}
+	client := &http.Client{
+		Transport: transport,
+	}
+	config := api.DefaultConfig()
+	if config.Error != nil {
+		return false, errwrap.Wrapf("failed to create api client: {{err}}", config.Error)
+	}
+	config.Address = leaderAddr
+	config.HttpClient = client
+	config.MaxRetries = 0
+	apiClient, err := api.NewClient(config)
+	if err != nil {
+		return false, errwrap.Wrapf("failed to create api client: {{err}}", err)
+	}
+
+	join := func() error {
+		// Unwrap the token
+		secret, err := apiClient.Logical().Write("sys/storage/raft/bootstrap/challenge", map[string]interface{}{
```

Comment:
I think we'll see some of the same issues here that we see with replication where if client certs are required this won't work as there's no way to provide them. May be a good thing for someone to take care of between now and final release, to add client cert handling in both places.

Reply:
Yeah this is a known TODO